### PR TITLE
Add AFSS training sampler

### LINKED ISSUE #24621: Anti-Forgetting Sampling Strategy (AFSS) for Faster YOLO Training
## Current status — 2026-09-06

**Optional, paused research; no core integration planned and no production blocker.** PR #26017 was [closed after a full review](https://github.com/ultralytics/ultralytics/pull/26017#issuecomment-5555295994). Its 544 net non-documentation lines spread sampling policy across shared dataset, augmentation, trainer, DDP, checkpoint, configuration, and validation paths. The maintenance cost is not justified by the available benefit evidence. No replacement implementation is requested.

The review also reproduced incorrect score attribution at `b32d24281c5044d32c55e0aa532eb40278bbbb56`: with the actual YOLO26n/COCO8 loader and validator, 2 of 4 training images received each other's precision/recall because the scoring dataset's aspect-ratio order differs from the training dataset's order. This is a local CPU correctness reproduction, not a performance benchmark or a current-main regression; AFSS has not been merged.

The paper's speedup claim and the reported slow 170k-image run on an unofficial branch remain historical evidence from different configurations. They do not establish a net benefit for this PR. Any future reconsideration would need both a substantially smaller design at the existing owner and representative comparisons reporting total wall time (including scoring), held-out accuracy, and supported distributed/resume behavior. A speedup on one workload alone would not justify this maintenance burden.

Earlier invitations to explore a minimal opt-in sampler remain below as historical context. The existing sampler is unchanged; this issue is retained for the research discussion rather than an active implementation commitment.

---

### Search before asking

- [x] I have searched the Ultralytics [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar feature requests.

### Description

AFSS (Anti-Forgetting Sampling Strategy) is a dynamic image sampling method proposed in a recent paper that reduces YOLO training time by skipping already-learned images each epoch, while preventing forgetting via periodic resampling.

It measures a per-image *learning sufficiency* score as `min(recall, precision)` and categorizes each image as **easy / medium / hard**, each with a tailored sampling policy:

| Level | Sampling policy |
|---|---|
| **Easy** | Sparsely resampled; priority to long-unseen images to prevent forgetting |
| **Medium** | Partially sampled; recently unseen images prioritized |
| **Hard** | Fully sampled every epoch |

### Use case

Training YOLO on large datasets (COCO-scale and above) is time-consuming because every image is processed every epoch, even those the model has already mastered.

AFSS would allow users to get **>1.43× training speedup** with no accuracy loss — or even slight gains — simply by replacing the default sampler.

The paper evaluates AFSS on YOLOv8, v10, v11 and v12 **using the Ultralytics framework directly**, across four benchmarks: MS COCO 2017, PASCAL VOC 2007, DOTA-v1.0 and DIOR-R.

📄 Paper: https://arxiv.org/abs/2603.17684
*Does YOLO Really Need to See Every Training Image in Every Epoch?* — Xie et al., March 2026

### Additional

The implementation would consist of a custom `Sampler` plugging into the existing `DataLoader` pipeline — non-invasive by design.

An optional training flag such as `sampler=afss` would be sufficient to expose the feature to users.

The authors conducted all their experiments directly on the Ultralytics codebase, which should make a future PR relatively straightforward.

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

### PR DESCRIPTION
**Pull request by Marchematics**
Adds AFSS adaptive sampling to YOLO training.

AFSS trains hard images every epoch, revisits learned images on a schedule, refreshes image scores, synchronizes DDP sampling, and resumes from checkpoints.

```python
model.train(data="coco.yaml", sampler="afss")
```

```bash
yolo train model=yolo26n.pt data=coco.yaml sampler=afss
```

Closes #24621

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @Marchematics, thank you for submitting a `ultralytics/ultralytics` 🚀 PR! This automated message confirms your contribution was received, and an Ultralytics engineer will assist with the review. To ensure a seamless integration of your work, please review the following checklist:

- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI) checks are passing. If any checks fail, please address the issues.
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com) for any new or modified features.
- ✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
- ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_ — Bruce Lee

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing). Don't hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀

**Comment by Marchematics** [comment]
I have read the CLA Document and I sign the CLA

File: ultralytics/models/yolo/detect/train.py [hunk 1]
```diff
@@ -59,7 +60,14 @@ def __init__(self, cfg=DEFAULT_CFG, overrides: dict[str, Any] | None = None, _ca
             overrides (dict, optional): Dictionary of parameter overrides for the default configuration.
             _callbacks (dict, optional): Dictionary of callback functions to be executed during training.
         """
+        overrides = dict(overrides or {})
+        sampler = overrides.pop("sampler", None)
```
Comment (UltralyticsAssistant) [inline ultralytics/models/yolo/detect/train.py:64]: ⚠️ **HIGH**: Removing `sampler` and `afss` before `BaseTrainer.__init__` drops them from `self.args`, but DDP workers are created from `vars(trainer.args)` and checkpoints save the same mapping as `train_args`. Consequently, `device='0,1'` AFSS runs launch child trainers without these flags and never register the AFSS callbacks; resuming without repeating `sampler=afss` also silently disables AFSS. Preserve the opt-in in trainer arguments or explicitly propagate it through DDP and checkpoint configuration.

File: ultralytics/utils/afss.py [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+            trainer.train_loader = trainer.get_dataloader(
+                trainer.data["train"],
+                batch_size=trainer.batch_size // trainer.world_size,
+                rank=LOCAL_RANK,
+                mode="train",
+                active_indices=selected,
+            )
+        else:
+            dataset = _unwrap_dataset(trainer.train_loader.dataset)
+            dataset.active_indices = selected
```
Comment (UltralyticsAssistant) [inline ultralytics/utils/afss.py:169]: ⚠️ **HIGH**: For `rect=True`, this changes the dataset length and index mapping but leaves `BaseDataset.batch` and `batch_shapes` computed for the full dataset. Because rectangular loaders disable shuffling, selected indices that skip across original batches can be combined into one active batch with different `rect_shape` values, causing `collate_fn`'s `torch.stack` to fail with a size mismatch. Recompute rectangular batch metadata for the active subset or reject/disable AFSS with rectangular training.

File: ultralytics/utils/afss.py [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+    validator(trainer)
+    if hasattr(loader, "close"):
+        loader.close()
+
+    if RANK in {-1, 0}:
+        dataset = _unwrap_dataset(loader.dataset)
+        index_by_name = {}
+        for index, filename in enumerate(dataset.im_files):
+            index_by_name.setdefault(str(filename), index)
+            index_by_name.setdefault(Path(filename).name, index)
```
Comment (UltralyticsAssistant) [inline ultralytics/utils/afss.py]: 💡 **MEDIUM**: The validator records image metrics under `Path(im_file).name`, so two images with the same basename in different directories overwrite one another. This basename alias also maps only to the first dataset entry, causing one image to receive another image's score while the other remains at its default hard score. Use a collision-free image identifier consistently for validation metrics and `index_by_name`.

File: ultralytics/utils/afss.py [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+    if trainer.world_size > 1:
+        payload = [trainer.afss_scheduler.state_dict() if RANK == 0 else None]
+        dist.broadcast_object_list(payload, src=0)
+        trainer.afss_scheduler.load_state_dict(payload[0])
+
+
+def afss_save_state(trainer):
+    """Save AFSS state alongside model checkpoints."""
+    if hasattr(trainer, "afss_scheduler") and RANK in {-1, 0}:
+        torch.save(trainer.afss_scheduler.state_dict(), trainer.wdir / "afss_state.pt")
```
Comment (UltralyticsAssistant) [inline ultralytics/utils/afss.py]: 💡 **MEDIUM**: All saves overwrite one `afss_state.pt`, independently of the checkpoint being written. If `best.pt` or an `epochN.pt` checkpoint is later resumed, this loader reads the state from the most recent save rather than the state corresponding to that checkpoint, so future precision/recall and `last_seen` values affect sampling for an older model. Store the scheduler state in each checkpoint or use checkpoint-specific sidecars and select the matching one on resume.

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

Reviewed all 10 changed files and the 413-line diff. AFSS has several concrete lifecycle and dataset-indexing issues: DDP and checkpoint configuration lose the opt-in, rectangular training can crash after subsetting, checkpoint sidecar state is not versioned, duplicate basenames corrupt per-image scores, and OOM recovery can bypass the selected subset. Not LGTM.

<details><summary>💬 Posted 5 inline comments</summary>

- ⚠️ **HIGH** `ultralytics/models/yolo/detect/train.py:64` Removing `sampler` and `afss` before `BaseTrainer.__init__` drops them from `self.args`, but DDP workers are created from `vars(trainer.args)` and checkpoints save the same mapping as `train_args`. Consequently, `device='0,1'` AFSS runs launch child trainers without these flags and never register the AFSS callbacks; resuming without repeating `sampler=afss` also silently disables AFSS. Preserve the opt-in in trainer arguments or explicitly propagate it through DDP and checkpoint configuration.
- ⚠️ **HIGH** `ultralytics/utils/afss.py:153` For `rect=True`, this changes the dataset length and index mapping but leaves `BaseDataset.batch` and `batch_shapes` computed for the full dataset. Because rectangular loaders disable shuffling, selected indices that skip across original batches can be combined into one active batch with different `rect_shape` values, causing `collate_fn`'s `torch.stack` to fail with a size mismatch. Recompute rectangular batch metadata for the active subset or reject/disable AFSS with rectangular training.
- 💡 **MEDIUM** `ultralytics/utils/afss.py:196` The validator records image metrics under `Path(im_file).name`, so two images with the same basename in different directories overwrite one another. This basename alias also maps only to the first dataset entry, causing one image to receive another image's score while the other remains at its default hard score. Use a collision-free image identifier consistently for validation metrics and `index_by_name`.
- 💡 **MEDIUM** `ultralytics/utils/afss.py:208` All saves overwrite one `afss_state.pt`, independently of the checkpoint being written. If `best.pt` or an `epochN.pt` checkpoint is later resumed, this loader reads the state from the most recent save rather than the state corresponding to that checkpoint, so future precision/recall and `last_seen` values affect sampling for an older model. Store the scheduler state in each checkpoint or use checkpoint-specific sidecars and select the matching one on resume.
- 📝 **LOW** `ultralytics/utils/afss.py:139` The equality check assumes the current loader still represents `trainer.afss_current_indices`. During BaseTrainer's auto-OOM recovery, `_build_train_pipeline()` rebuilds a full-data loader but does not update this field; on the retry, the same selected list compares equal and is not reapplied. A resumed run that OOMs on its first resumed epoch can therefore train on the full dataset while logging the AFSS subset, and remain unsampled until the plan changes. Reapply the active indices whenever t…
</details>

File: ultralytics/utils/afss.py [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+        if trainer.epochs - epoch <= trainer.afss_scheduler.FULL_FINAL_EPOCHS
+        else trainer.afss_scheduler.sample_indices(epoch)
+    )
+
+    if trainer.world_size > 1:
+        payload = [selected if RANK == 0 else None]
+        dist.broadcast_object_list(payload, src=0)
+        selected = payload[0]
+
+    if selected != trainer.afss_current_indices:
```
Comment (UltralyticsAssistant) [inline ultralytics/utils/afss.py:155]: 📝 **LOW**: The equality check assumes the current loader still represents `trainer.afss_current_indices`. During BaseTrainer's auto-OOM recovery, `_build_train_pipeline()` rebuilds a full-data loader but does not update this field; on the retry, the same selected list compares equal and is not reapplied. A resumed run that OOMs on its first resumed epoch can therefore train on the full dataset while logging the AFSS subset, and remain unsampled until the plan changes. Reapply the active indices whenever the rebuilt loader does not reflect them.

**Comment by Marchematics** [comment]
All review comments are fixed in the latest commits. CI is green.

**Comment by pderrenger** [comment]
Thanks for the update. I’ve confirmed all current CI checks pass; we’ll re-review the latest commits against the previously raised concerns.

**Comment by glenn-jocher** [comment]
Thank you @Marchematics for contributing this implementation and working through the earlier review feedback. After independently reviewing all 12 changed files at b32d24281c5044d32c55e0aa532eb40278bbbb56 and the discussion in #24621, we are closing this PR without merging.

The maintenance cost is too high for the demonstrated benefit. This adds 544 net non-documentation lines, including 250 net test lines and a new 217-line scheduling module. It changes shared dataset indexing, mosaic selection, trainer iteration/warmup bookkeeping, DDP loader rebuilding, checkpoint serialization, configuration acceptance, and validation metric keys. It also retains a legacy sidecar compatibility path for an unshipped feature, while several tests replace the actual trainer, loader, or validator with fakes. This exceeds the scope of the minimal opt-in sampler previously discussed in #24621.

There is also a confirmed correctness problem in the current head. `afss_refresh_metrics()` builds its filename-to-index mapping from the scoring loader, whose validation dataset is sorted by aspect ratio. Those positions are then used in the scheduler arrays that index the original, normally non-rectangular training dataset. Full-path keys fix basename collisions but do not fix this ordering mismatch.

Local reproduction: Python 3.12.12, PyTorch 2.14.0, CPU, pretrained YOLO26n, the four official COCO8 training images, `imgsz=160`, `batch=2`, `workers=0`, `rect=False`, `sampler="afss"`, `amp=False`. Construct the actual `DetectionTrainer`, run `_setup_train()`, initialize AFSS at epoch 3 of 20, run a real training forward pass to initialize loss items, then call the actual training-set validator and `afss_refresh_metrics()`. No mocked implementation or fabricated scores were used. Two of four images receive another image's scores:

| Training image | Correct precision / recall | Stored AFSS precision / recall | Actual source of stored scores |
| --- | --- | --- | --- |
| `000000000009.jpg` | 0.048 / 0.75 | 0.200 / 1.00 | `000000000034.jpg` |
| `000000000034.jpg` | 0.200 / 1.00 | 0.048 / 0.75 | `000000000009.jpg` |

This is a correctness reproduction, not a speed benchmark. The PR and linked discussion provide no matched end-to-end wall-time and held-out accuracy comparison for this head, including its repeated full-training-set scoring passes. The paper's speedup claim and the older slowdown report about an unofficial `afss-new` branch do not establish this implementation's performance. Green CI and selecting fewer images are insufficient evidence of a net improvement.

Our decision is **no core integration and no replacement work requested**. Fixing the attribution bug, or showing a speedup on one workload, would not by itself justify maintaining this cross-cutting feature. The existing sampler remains unchanged. #24621 retains the historical discussion as optional, paused research rather than a production blocker. Thanks again for the time and effort you put into the contribution.
